### PR TITLE
Support grouped execution for eligible table scans

### PR DESCRIPTION
**Pull request by shixuan-fan**
This is controlled by session property grouped_execution_for_all_eligible_table_scans.

Related to https://github.com/prestodb/presto/issues/12124

### CODE REVIEW

**Comment by rschlussel** [comment]
So is this to say that we should always do grouped execution for eligible tablescans even if the nodes above it don't care about grouped execution?  Can you give more context about why/when we want this behavior?

**Comment by wenleix** [comment]
@rschlussel : This allows partial recovery for ScanFilterProject-only queries (reading from bucketed table).

Also, today reading from bucketed table with ungrouped execution is somewhat hacky, for example, it requires (https://github.com/prestodb/presto/wiki/HiveSplitSource-and-Grouped-Execution#old-days):

* the underlying connector to return the split in a round-robin fashion 
* A special `BucketedSplitPlacmentPolicy` to be used, which is still aware of bucketed table. 

So popularizing grouped execution seems to be the right direction in the long-term -- i.e. the stage scheduler is aware that he is scheduling split in a bucket-aware way.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -212,6 +213,11 @@ public SystemSessionProperties(
                         "Use grouped execution for aggregation when possible",
                         featuresConfig.isGroupedExecutionForAggregationEnabled(),
                         false),
+                booleanProperty(
+                        GROUPED_EXECUTION_FOR_ALL_ELIGIBLE_TABLE_SCANS,
+                        "Use grouped execution for all eligible table scans",
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: nit: `Experimental: use grouped execution for all eligible table scans` :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 2]
```diff
@@ -1059,11 +1060,12 @@ public GroupedExecutionProperties visitTableScan(TableScanNode node, Void contex
             }
             List<ConnectorPartitionHandle> partitionHandles = nodePartitioningManager.listPartitionHandles(session, tablePartitioning.get().getPartitioningHandle());
             boolean recoveryEligible = metadata.getConnectorCapabilities(session, node.getTable().getConnectorId()).contains(SUPPORTS_REWINDABLE_SPLIT_SOURCE);
+            boolean useful = isGroupedExecutionForAllEligibleTableScansEnabled(session);
             if (ImmutableList.of(NOT_PARTITIONED).equals(partitionHandles)) {
-                return new GroupedExecutionProperties(false, false, ImmutableList.of(), 1, recoveryEligible);
+                return new GroupedExecutionProperties(false, useful, ImmutableList.of(), 1, recoveryEligible);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: I feel `GroupedExecutionProperties#useful` should be renamed. But don't have to do it in this PR.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 3]
```diff
@@ -332,6 +333,19 @@ public FeaturesConfig setGroupedExecutionForAggregationEnabled(boolean groupedEx
         return this;
     }
 
+    public boolean isGroupedExecutionForAllEligibleTableScansEnabled()
+    {
+        return groupedExecutionForAllEligibleTableScansEnabled;
+    }
+
+    @Config("grouped-execution-for-all-eligible-table-scans-enabled")
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: any thought about the name @rschlussel ? -- do we need "all" from grammar perspective?  😃

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 4]
```diff
@@ -2884,6 +2886,12 @@ private void testGroupedExecution(Session session)
                     .setSystemProperty(DYNAMIC_SCHEDULE_FOR_GROUPED_EXECUTION, "true")
                     .build();
 
+            // All eligible table scans only, 1 group per worker at a time
+            Session allEligibleTableScans = Session.builder(session)
+                    .setSystemProperty(GROUPED_EXECUTION_FOR_ALL_ELIGIBLE_TABLE_SCANS, "true")
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I think we can also enable this in previous tests as well.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 3 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: i don't think we need 'all'.  it does emphasize that we'll do grouped execution even if there's no aggregation/join on top, but I still think it's better without 'all' to make it shorter.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: if it's experimental, should the config property be prefixed with `experimental.`?

**Pull request merged**